packages/core/src/service/services/api-key.service.ts

```diff
@@ -0,0 +1,380 @@
+        private translator: TranslatorService,
+        private userService: UserService,
+    ) {}
+
+    /**
+     * @description
+     * Returns the appropriate {@link ApiKeyStrategy} based on the {@link ApiType}.
+     * This is needed because the admin and shop ApiKeyStrategy may differ.
+     */
+    getApiKeyStrategyByApiType(apiType: ApiType): ApiKeyStrategy {
+        return apiType === 'admin'
+            ? this.configService.authOptions.adminApiKeyStrategy
+            : this.configService.authOptions.shopApiKeyStrategy;
+    }
+
+    /**
+     * @description
+     * Checks that the active user is allowed to grant the specified Roles for an API-Key
+     *
+     * // TODO this is taken & slightly modified from adminservice, could merge to not repeat logic
+     *
+     * @throws {UserInputError} If the active User has insufficient permissions
+     * @returns Role-Entities with relations to Channels
+     */
+    private async assertActiveUserCanGrantRoles(ctx: RequestContext, roleIds: ID[]): Promise<Role[]> {
+        if (roleIds.length === 0) return [];
+
+        const roles = await this.connection.getRepository(ctx, Role).find({
+            where: { id: In(roleIds) },
+            relations: { channels: true },
+        });
+        const permissionsRequired = getChannelPermissions(roles);
+        for (const channelPermissions of permissionsRequired) {
+            const isAllowed = await this.roleService.userHasAllPermissionsOnChannel(
+                ctx,
+                channelPermissions.id,
+                channelPermissions.permissions,
+            );
+
+            if (!isAllowed)
+                throw new UserInputError('error.active-user-does-not-have-sufficient-permissions');
+        }
+
+        return roles;
+    }
+
+    /**
+     * @description
+     * Simple user identifier generation function because it is a non-nullable field on User.
+     *
+     * Because this simply appends the lookupId, some databases like MySQL/Maria may run into
+     * length issues if the lookupId has too many characters. Practically speaking this
+     * should not happen but worth to keep in mind.
+     *
+     * @internal
+     */
+    private generateApiKeyUserIdentifier(lookupId: string): string {
+        return `apikey-user-${lookupId}`;
+    }
+
+    /**
+     * @description
+     * Creates a new API-Key for the given User
+     *
+     * **Important**: The caller is responsible for avoiding privilege escalations by
+     * verifying `userIdOwner` and `userIdApiKeyUser`; **Use this with great care!**
+     *
+     * If you allow users to specify these IDs, they may leak existing User IDs via thrown errors.
+     *
+     * @throws {EntityNotFoundError} When either Owner or ApiKeyUser cannot be found
+     * @throws {UserInputError} When the User tries to grant a role which they themselves dont have
+     */
+    async create(
+        ctx: RequestContext,
+        input: CreateApiKeyInput,
+        userIdOwner: ID,
```

Comment:
Maybe a little nitpicky, but I had some trouble understanding the methods description in combination with the `userIdOwner` param. Would it make sense to name this something like `callerUserId`, so that it is clear that we need the callers userId to check for privilege escalation?

Reply:
The owner basically only becomes relevant for plugins where customers can own a key, so I think in line with the `Permission.Owner` - Owner is clearer than caller.